File: bin/tagify.rb
#!/usr/bin/env ruby

require 'csv'
require 'set'
require 'unicode'

flickr_tags = Set.new
tag_source = File.expand_path('../../data/tags_from_flickr.csv', __FILE__)
CSV.foreach(tag_source) do |row|
  flickr_tags.add row[0]
end

source_file = File.expand_path('../../data/POPTagList_Mitch_and_Laura_Categories_Sheet1.csv', __FILE__)

outfile = File.expand_path('../../data/updated_tags.csv',__FILE__)
CSV.open(outfile, 'w+') do |csv|
  CSV.foreach(source_file) do |row|
    text, raw, mitch, laura, model, field, basis = row
    default = 'NO_MATCH'
    if text.nil? || text.strip() == ''
      tag = Unicode::normalize_C(raw.downcase).gsub(/[^[:word:]_]/,'')
      row[0] = flickr_tags.include?(tag) ? tag : "NEW:NO_MATCH:#{tag}"
    else
      if ! flickr_tags.include? text
        row[0] =  "CURR:NO_MATCH:#{text}"
      end
    end
    csv << row
  end
end

puts "Wrote: #{outfile}"

File: bin/tagify.py
# -*- coding: utf-8 -*-
import os
import sys
import csv
import re
import unicodedata

sys.path.append(os.path.dirname(__file__))
from unicode_csv import UTF8Recoder, UnicodeReader, UnicodeWriter

rdr = UnicodeReader(open(sys.argv[1]))
tag_re = re.compile('\W', re.UNICODE)
clean_re = re.compile('[\s"!@#\$%^&*():_+=\'/.;`<>\[\]?\\-]')

# clean_tag = raw_tag.gsub(/[\\s"!@#\\$\\%^&*():\\-_+=\\'\\/.;`<>\\[\\]?\\\\]/,"").downcase


def strip_accents(s):
    return ''.join(c for c in unicodedata.normalize('NFD', s)
                   if unicodedata.category(c) != 'Mn')


outfile = open('updated_tags.orig.csv', 'w+')

wrtr = UnicodeWriter(outfile, delimiter='\t')
head = [ 'Normalized', 'Raw', 'Model', 'field' ]
wrtr.writerow(head)

for row in rdr:
    new_row = []
    ugly, raw, mitch, laura, model, field, basis = row[:]
    new_row = [ ugly, raw, model, field ]

    if ugly:
        new_row = [ ugly, raw, model, field ]
    else:
        u = unicodedata.normalize('NFC', raw)
        u = tag_re.sub('', u.strip().lower())
        new_row = [ u, raw, model, field ]
    wrtr.writerow(new_row)

outfile.close()
print "Wrote \"%s\"" % (outfile.name, )
